pywayland/utils.py

```diff
@@ -76,3 +78,49 @@ def close(self):
 
         os.close(self.fd)
         self.fd = None
+
+
+def wl_container_of(ptr: ffi.CData, ctype: str, member: str, ffi=ffi) -> ffi.CData:
+    """
+    #define wl_container_of(ptr, sample, member)				\
+            (__typeof__(sample))((char *)(ptr) -				\
+                                 offsetof(__typeof__(*sample), member))
+
+    :param ptr:
+        Pointer to contained object
+    :param ctype:
+        ctype of container as string
+    :param member:
+        Member name of contained object in ctype
+    :param ffi:
+        ffi module to use. The default is pywayland, but this allows the use of this
+        macro by other ffi modules that use `wl_list`s.
+    """
+    return ffi.cast(
+        ctype,
+        ffi.cast("char *", ptr) - ffi.offsetof(ctype, member)
```

Comment:
At least in `pywayland/server/listener.py`, Used `f"{ctype} *"` for the first ctype and leave `ffi.offsetof(ctype, member)` for the second ctype. Not sure if that is needed or if cffi is smart enough here.